packages/ra-ui-materialui/src/input/SelectArrayInput.tsx

```diff
@@ -159,6 +159,16 @@ export const SelectArrayInput = (props: SelectArrayInputProps) => {
             // We might receive an event from the mui component
             // In this case, it will be the choice id
             if (eventOrChoice?.target) {
+                eventOrChoice.target.value = eventOrChoice.target.value.reduce(
+                    (acc, value) => {
+                        // eslint-disable-next-line eqeqeq
+                        const index = acc.findIndex(v => v == value);
+                        return index < 0
+                            ? [...acc, value]
+                            : [...acc.slice(0, index), ...acc.slice(index + 1)];
+                    },
+                    []
+                );
```

Comment:
what the heck is this supposed to do??

At least it requires some comments, but I even fear it's wrong (why do you remove the value if it was found in the accumulator?)

Reply:
`eventOrChoice.target.value` can, in this case, be set in this case as `[2, '2', 1, '1']`

Reply:
Added comment.

Reply:
If this case can happen then it should be covered by a test.

I'm not convinced it can happen but event if it did, I tested it manually by changing the 'DifferentIdTypes' story, but this code doesn't work. It doesn't remove duplicates, and worse, it removes the wrong item.

Reply:
Okay, took me a while to actually understand why you need this snippet and how it works.
Now I am convinced that it works and you should keep it.

I'd simply recommend being a little more verbose in the comments because it's really not straightforward:
```js
                // when used with different IDs types, unselection leads to double selection with both types
                // instead of the value being removed from the array
                // e.g. we receive eventOrChoice.target.value = [1, '2', 2] instead of [1] after removing 2
                // this snippet removes a value if it is present twice
```

And it is indeed covered by the test `should unselect values when ids type are different`, so no worries there, my bad!